library/alloc/src/boxed/thin.rs

```diff
@@ -0,0 +1,219 @@
+impl<T: ?Sized + Debug> Debug for ThinBox<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Debug::fmt(self.deref(), f)
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized + Display> Display for ThinBox<T> {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        Display::fmt(self.deref(), f)
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> Deref for ThinBox<T> {
+    type Target = T;
+
+    fn deref(&self) -> &T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts(value as *const (), metadata);
+        unsafe { &*pointer }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> DerefMut for ThinBox<T> {
+    fn deref_mut(&mut self) -> &mut T {
+        let value = self.data();
+        let metadata = self.meta();
+        let pointer = ptr::from_raw_parts_mut::<T>(value as *mut (), metadata);
+        unsafe { &mut *pointer }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> Drop for ThinBox<T> {
+    fn drop(&mut self) {
+        unsafe {
+            // Vendor the impl of DerefMut to help the borrow checker see that borrows are
+            // separate
+            let value = {
+                let value = self.data();
+                let metadata = self.meta();
+                let pointer = ptr::from_raw_parts_mut::<T>(value as *mut (), metadata);
+                &mut *pointer
+            };
+            self.ptr.drop::<T>(value);
+        }
+    }
+}
+
+#[unstable(feature = "thin_box", issue = "92791")]
+impl<T: ?Sized> ThinBox<T> {
+    fn meta(&self) -> <T as Pointee>::Metadata {
+        //  Safety:
+        //  -   NonNull and valid.
+        unsafe { *self.ptr.header() }
+    }
+
+    fn data(&self) -> *mut u8 {
+        self.ptr.value()
+    }
+}
+
+struct WithHeader<H>(NonNull<u8>, PhantomData<H>);
+
+impl<H> WithHeader<H> {
+    #[cfg(not(no_global_oom_handling))]
+    fn new<T>(header: H, value: T) -> WithHeader<H> {
+        let layout = Self::alloc_layout(mem::size_of::<T>(), mem::align_of::<T>());
+        let aligned_header_size = Self::aligned_header_size(layout.align());
+
+        unsafe {
+            let ptr = alloc::alloc(layout);
+
+            if ptr.is_null() {
+                alloc::handle_alloc_error(layout);
+            }
+            //  Safety:
+            //  -   The size is at least `aligned_header_size`.
+            let ptr = ptr.offset(aligned_header_size as isize) as *mut _;
+
+            let ptr = NonNull::new_unchecked(ptr);
+
+            let result = WithHeader(ptr, PhantomData);
+            ptr::write(result.header(), header);
+            ptr::write(result.value().cast(), value);
+
+            result
+        }
+    }
+
+    //  Safety:
+    //  -   Assumes that `value` can be dereferenced.
+    unsafe fn drop<T: ?Sized>(&self, value: &mut T) {
```

Comment:
Isn't this risky? I remember there being issues in the past with an (argument) reference's backing storage being deallocated mid-way through a function. Something about the (argument) reference being guaranteed alive at the start vs "throughout". Might need someone else to explain it though.

Maybe running this in miri, with stacked borrows enabled, might be enough to detect an issue? Not sure though, the raw pointers might mask it partly.

Simplest change AFAICT (if this is a problem) would be to make it `value: *mut T` instead, and turn it into a reference only for `{size,align}_of_val`.

Reply:
I updated it from `value: *mut T` because of [this comment](https://github.com/rust-lang/rust/pull/90066#discussion_r783283147) by @a1phyr. I'm not really sure who is right or what to do so I'm gonna also cc @matthieu-m who originally wrote the code and did use `*mut T` so they can weigh in as well since this is one of the last unresolved comments atm.

Reply:
I think #55005 is an example of what @eddyb means.

Reply:
Okay, this should be resolved now.

Reply:
Yeah using a raw ptr seems better.